### PR TITLE
Overlay driver to support network layer encryption

### PR DESCRIPTION
**Pull request by aboch**
- Allows network layer encryption for overlay driver

Signed-off-by: Alessandro Boch <aboch@docker.com>

### CODE REVIEW

**Comment by aboch** [comment]
How to test it:

1) On each node in the cluster set the encryption (key,tag) pair with:
```
docker network create -d overlay --opt key=d0562776bf0e75830ba3f7f8eb6c09b537b4e89ccf7097445479bc06c7b2b6e5 --opt tag=78917750 <any name>
```

2) Create one or more secure overlay network along with a regular, non secure network:
```
docker network create -d overlay --opt secure sec
docker network create -d overlay nosec
```

3) Verify containers running on secure network(s) on different hosts can talk to each other. The secure network traffic across hosts will happen via esp packets (ip.proto==50). Verify traffic on non-secure overlay network is still fine and happening via plain vxlan packets:

In this example there are two secure and one non-secure overlay network, three nodes (192.168.201.54, 192.168.203.151, 192.168.202.224) each with one container on each network pinging contaienrs on different hosts
2016-06-01-111532_1920x1080_scrot


4) Rotate the encryption key on each host:
```
docker network create -d overlay --opt newKey=11112776bf0e75830ba3f7f8eb6c09b537b4e89ccf7097445479bc06c7b21111 --opt newTag=54321234 <any name>
```
During the rotate process you should not see significant traffic drops. Hosts will be able to receive on old and new key until peers has completed the switch to the new key.

File: agent.go [hunk 1]
```diff
@@ -86,6 +90,10 @@ func (c *controller) handleKeyChange(keys []*types.EncryptionKey) error {
 			if key.Subsystem == "networking:gossip" {
 				a.networkDB.SetKey(key.Key)
 			}
+			if key.Subsystem == "networking:gossip" /*"networking:ipsec"*/ {
+				drvEnc.Key = hex.EncodeToString(key.Key)
```
Comment (sanimej) [inline agent.go]: keys should be []byte. Its unnecessary to covert it here and back to byte slice in the driver code.

File: agent.go [hunk 2]
```diff
@@ -86,6 +90,10 @@ func (c *controller) handleKeyChange(keys []*types.EncryptionKey) error {
...
+				drvEnc.Tag = strconv.FormatUint(key.LamportTime, 10)
+			}
```
Comment (sanimej) [inline agent.go]: LamportTime can also be passed as a number rather than a string.

File: drivers/overlay/joinleave.go [hunk 3]
```diff
@@ -27,6 +27,10 @@ func (d *driver) Join(nid, eid string, sboxKey string, jinfo driverapi.JoinInfo,
 		return fmt.Errorf("could not find endpoint with id %s", eid)
 	}
 
+	if n.secure && len(d.keys) == 0 {
+		return fmt.Errorf("cannot join secure network: encryption keys not present")
+	}
+
```
Comment (sanimej) [inline drivers/overlay/joinleave.go]: In the non-cluster mode may be its better to error out in libnetwork on network creation rather than join ?

**Comment by mavenugo** [comment]
LGTM

File: agent.go (hunk 1 above)
Comment (aboch) [inline agent.go]: I kept the logic we have in discoverapi types of having everything as string or []string.
But yeah, we can probably break that tradition and avoid this unnecessary double conversion.
If you do not mind, I'd address this as part of future bug fixes, taking some time to assess the consequences, it is anyway only affecting internal APIs.

**Comment by sanimej** [comment]
@aboch We can also address the comments in a bug fix.

File: agent.go (hunk 2 above)
Comment (aboch) [inline agent.go]: Agree. Same as above reply.

**Comment by pierreozoux** [comment]
This looks good :)
How can I test? what are the keys and tags?
When can we expect documentation to ship?

Thanks!

**Comment by sanimej** [comment]
@pierreozoux You can try it out with the swarm mode using the 1.12 RC3 docker binary. When creating the overlay network pass the option `--opt secure`. Key allocation/rotation etc. are in-built in the docker swarm mode.

**Comment by pierreozoux** [comment]
Ok, thanks, I'll try it!

On 04-07-2016 08:47, Santhosh Manohar wrote:
> @pierreozoux <https://github.com/pierreozoux> You can try it out with
> the swarm mode using the 1.12 RC3 docker binary. When creating the
> overlay network pass the option |--opt secure|. Key allocation/rotation
> etc. are in-built in the docker swarm mode.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/libnetwork/pull/1199#issuecomment-230226724>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABxvHXUeBfD2oCj6tbM3mgvlCHJvsfVXks5qSLqngaJpZM4Iq5kQ>.
> 

-- 
I use PGP to protect our privacy, if you want to know more, you can
follow this
https://emailselfdefense.fsf.org/en/

If you have further questions, please do not hesitate to ask.
You can verify my public key here: https://keybase.io/pierreozoux

**Comment by kalahari** [comment]
Is there any documentation or blog posts on the encryption mode or algorithms this implements?

**Comment by aboch** [comment]
@kalahari Docker documentation is being updated on how the dataplane encryption can be turned on, but I am not sure it will contain the details you are asking.

So this implements autheticated encryption ipsec in host-to-host transport mode.
The algorithm used is the one specified in rfc4106, 128 bits key length.

**Pull request merged**